src/algorithms/tonal/pitchyinprobabilities.cpp

```diff
@@ -0,0 +1,333 @@
+
+  size_t minTau = 2;
+  size_t maxTau = _yin.size();
+
+  Real minWeight = 0.01;
+  size_t nThreshold = 100;
+  int nThresholdInt = nThreshold;
+
+  vector<Real> thresholds(nThresholdInt);
+  vector<Real> distribution(nThresholdInt);
+  _peakProb.assign(_yin.size(), 0.0);
+
+  for (int i = 0; i < nThresholdInt; ++i) {
+    distribution[i] = betaDist2[i];
+    thresholds[i] = 0.01 + i*0.01;
+  }
+
+  int currThreshInd = nThreshold-1;
+  size_t tau;
+  tau = minTau;
+  size_t minInd = 0;
+  Real minVal = 42.0;
+  Real sumProb = 0.0;
+  while (tau+1 < maxTau) {
+    if (_yin[tau] < thresholds[thresholds.size()-1] && _yin[tau+1] < _yin[tau]) {
+      while (tau + 1 < maxTau && _yin[tau+1] < _yin[tau]) {
+        tau++;
+      }
+
+      if (_yin[tau] < minVal && tau > 2) {
+        minVal = _yin[tau];
+        minInd = tau;
+      }
+
+      currThreshInd = nThresholdInt-1;
+      while (thresholds[currThreshInd] > _yin[tau] && currThreshInd > -1) {
+        _peakProb[tau] += distribution[currThreshInd];
+        currThreshInd--;
+      }
+      sumProb += _peakProb[tau];
+      tau++;
+    } else {
+      tau++;
+    }
+  }
+
+  if (_peakProb[minInd] > 1) {
+    E_WARNING("WARNING: yin has prob > 1 ??? I'm returning all zeros instead.");
+    _peakProb = vector<Real>(_yin.size());
+  }
+    
+  Real nonPeakProb = 1.0;
+  if (sumProb > 0) {
+    for (int i = minTau; i < maxTau; ++i)
+    {
+        _peakProb[i] = _peakProb[i] / sumProb * _peakProb[minInd];
+        nonPeakProb -= _peakProb[i];
+    }
+  }
+
+  if (minInd > 0) {
+    _peakProb[minInd] += nonPeakProb * minWeight;
+  }
+
+  RMS = sqrt(sumSquare(signal, 0, _yin.size()) / _yin.size());
+
+  // reuse the vector
+  _freq.resize(0);
+  _peakProb_freq.resize(0);
+
+  // calculate frequency probabilities
+  for (size_t iBuf = 0; iBuf < _yin.size(); ++iBuf) {
+    if (_peakProb[iBuf] > 0) {
+      Real currentF0 = _sampleRate * (1.0 / parabolicInterpolation(_yin, iBuf, _yin.size()));
+      _freq.push_back(currentF0);
+      _peakProb_freq.push_back(_peakProb[iBuf]);
+    }
+  }
+
+  // convert the frequency to cents
+  bool isLowAmplitude = (RMS < _lowAmp);
+
+  for (size_t iCandidate = 0; iCandidate < _freq.size(); ++iCandidate) {
+    Real pitchCents = 12 * log(_freq[iCandidate]/440)/log(2.) + 69;
+    _freq[iCandidate] = pitchCents;
+    if (isLowAmplitude) {
+      Real factor = ((RMS+0.01 * _lowAmp) / (1.01 * _lowAmp));
```

Comment:
Explain why adding 0.01

Reply:
it's a part of the algorithm. I don't think we need to explain this, or we need to explain for each step of the algorithm in the code.

Reply:
In this particular case, those are some hardcoded values (0.01, 1.01), so that's good to have a comment that we took those from the paper and what's their sense.

Reply:
I don't totally agree, but since you are the boss. And it's more simple to just add the comment.